port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,433 @@
+
+typedef struct frameData {
+	void **address_array;
+	uintptr_t capacity;
+} frameData;
+
+/**
+ * Handles segfaults where either an invalid address is called or a signal was received in kernel mode.
+ * Both cases leave the ip register address invalid and libunwind fails to iterate past _sigtramp. This
+ * function fetches previous frame information and restores correct register values to the unwind cursor.
+ * It also records the stack frame addresses of the remaining frames.
+ *
+ * @param[in] cursor keeps track of stack unwind state
+ * @param[out] array contains the backtrace addresses to be returned
+ * @param[in] size size of the array
+ * @param[in] nAddr number of addresses currently recorded in the array
+ *
+ * @return total number of addresses recorded in the array
+ */
+static uintptr_t
+handle_sigtramp(unw_cursor_t *cursor, void **array, uintptr_t size, uintptr_t nAddr)
+{
+	unw_word_t prevFrame = 0;
+	unw_word_t ip = 0;
+	ucontext_t *uc = NULL;
+	OMRUnixSignalInfo info;
+	const char *name = NULL;
+	void *reg = NULL;
+
+#if defined(OMR_ARCH_X86)
+	/* ucontext for frame prior to _sigtramp is stored in rbx. Fetch previous frame info from it.
+	 * See https://opensource.apple.com/source/Libc/Libc-498/x86_64/sys/_sigtramp.s.auto.html
+	 */
+	unw_get_reg(cursor, UNW_X86_64_RBX, &prevFrame);
+	uc = (ucontext_t *)prevFrame;
+	info.platformSignalInfo.context = uc;
+
+	/* Copy the necessary machine regs to cursor so we can iterate the previous frame.
+	 * See https://opensource.apple.com/source/libunwind/libunwind-35.4/src/unw_getcontext.s.auto.html
+	 */
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_BP, &name, &reg);
+	unw_set_reg(cursor, UNW_X86_64_RBP, *(unw_word_t *)reg);
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_SP, &name, &reg);
+	/* Store call-site sp (skip return address) into RSP. */
+	unw_set_reg(cursor, UNW_X86_64_RSP, (*(unw_word_t *)reg) + 8);
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_PC, &name, &reg);
+	ip = *(unw_word_t *)reg;
+
+	if ((0 != size) && (nAddr < (size - 1))) {
+		/* libunwind failed to iterate past _sigtramp because of a call to invalid address or a signal
+		 * received in kernel (right after syscall). In either case, the return address will be
+		 * pointed to by rsp: in the first case, return address pushed to stack right before call,
+		 * in the second case, syscall wrapper's caller pushed its return address. Record the ip and
+		 * skip the frame.
+		 */
+		array[nAddr] = (void *)ip;
+		nAddr += 1;
+		infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_SP, &name, &reg);
+		ip = **(unw_word_t **)reg;
+
+		/* Record the return address. */
+		array[nAddr] = (void *)ip;
+		nAddr += 1;
+		unw_set_reg(cursor, UNW_REG_IP, ip);
+
+		while ((unw_step(cursor) > 0) && (nAddr < size)) {
+			unw_get_reg(cursor, UNW_REG_IP, &ip);
+			array[nAddr] = (void *)ip;
+			nAddr += 1;
+		}
+	}
+#endif /* defined(OMR_ARCH_X86) */
```

Comment:
Please add the following code here until I (or someone else) implement the function for AArch64:
```
#else /* defined(OMR_ARCH_X86) */
        nAddr = 0;
```

Reply:
Currently for AArch64, this function will return the number of addresses already in the array (i.e. the number of signal handler frames). If I return 0 instead, then the signal handler frames recorded in the addresses array will be ignored and the javacore will display `no native callstack` for the non-current threads, is this the intention?

Reply:
Please ignore my previous comment.  I tried this again today, and it seems to work without changing.